File: app/src/main/java/com/example/policyfolio/UI/Nominee/AddNomineeFragment.java
package com.example.policyfolio.UI.Nominee;


import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;

import android.telephony.PhoneNumberFormattingTextWatcher;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.example.policyfolio.R;
import com.example.policyfolio.UI.Adapters.ListAdapters.BasicDropdownTextAdapter;
import com.example.policyfolio.Util.Constants;
import com.example.policyfolio.ViewModels.WithUser.NomineeViewModel;

/**
 * A simple {@link Fragment} subclass.
 */
public class AddNomineeFragment extends Fragment implements BasicDropdownTextAdapter.ParentCallback {

    private View rootView;
    private NomineeViewModel viewModel;

    private EditText name;
    private EditText email;
    private EditText phone;
    private EditText alternativePhone;

    private TextView relationText;
    private LinearLayout relation;
    private BasicDropdownTextAdapter textAdapter;
    private String[] relations;

    private Button done;

    private NomineeCallback callback;

    public AddNomineeFragment() {
        // Required empty public constructor
    }

    public AddNomineeFragment(NomineeCallback callback){
        this.callback = callback;
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_add_nominee, container, false);
        viewModel = ViewModelProviders.of(getActivity()).get(NomineeViewModel.class);

        name = rootView.findViewById(R.id.name);
        email = rootView.findViewById(R.id.email);
        phone = rootView.findViewById(R.id.phone);
        alternativePhone = rootView.findViewById(R.id.alt_phone);

        relationText = rootView.findViewById(R.id.relation_text);
        relation = rootView.findViewById(R.id.relation);

        relations = getResources().getStringArray(R.array.relationship_array);
        textAdapter = new BasicDropdownTextAdapter(getContext(),relations,this, Constants.ListTypes.RELATIONSHIPS, getResources().getColor(R.color.colorPrimaryDarkest));

        done = rootView.findViewById(R.id.done);

        setListeners();

        return rootView;
    }

    private void setListeners() {
        phone.addTextChangedListener(new PhoneNumberFormattingTextWatcher());
        alternativePhone.addTextChangedListener(new PhoneNumberFormattingTextWatcher());
        name.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                String string = s.toString();
                if(string.length()>0)
                    name.setBackgroundColor(getResources().getColor(android.R.color.transparent));
                else
                    name.setBackground(getResources().getDrawable(R.drawable.text_background_grey_8dp));
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        email.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                String string = s.toString();
                if(string.length()>0)
                    email.setBackgroundColor(getResources().getColor(android.R.color.transparent));
                else
                    email.setBackground(getResources().getDrawable(R.drawable.text_background_grey_8dp));
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        phone.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                String string = s.toString();
                if(string.length()>0)
                    phone.setBackgroundColor(getResources().getColor(android.R.color.transparent));
                else
                    phone.setBackground(getResources().getDrawable(R.drawable.text_background_grey_8dp));
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        alternativePhone.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                String string = s.toString();
                if(string.length()>0)
                    alternativePhone.setBackgroundColor(getResources().getColor(android.R.color.transparent));
                else
                    alternativePhone.setBackground(getResources().getDrawable(R.drawable.text_background_grey_8dp));
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        viewModel.setRelation(0);
        relation.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                callback.openListSheet(Constants.ListTypes.RELATIONSHIPS,textAdapter);
            }
        });

        done.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String e = email.getText().toString();
                String n = name.getText().toString();
                String p = phone.getText().toString();
                String aP = alternativePhone.getText().toString();

                Log.e("DETAILS",e + " " + p + " " + n + " " + aP);

                if(!android.util.Patterns.EMAIL_ADDRESS.matcher(e).matches())
                    callback.showSnackbar("Enter a valid email address");
                else if(email.length()>0 && phone.length()>0 && name.length()>0 && relationText.getCurrentTextColor()==getResources().getColor(R.color.colorPrimaryDark)){
                    viewModel.setEmail(e);
                    viewModel.setAlternateNumber(aP);
                    viewModel.setName(n);
                    viewModel.setPhone(p);
                    callback.done();
                }
                else {
                    callback.showSnackbar("Incomplete Information");
                }
            }
        });
    }

    @Override
    public void setValue(int position, int type) {
        callback.closeListSheet();
        if(type == Constants.ListTypes.RELATIONSHIPS){
            relationText.setText(relations[position]);
            relationText.setTextColor(getResources().getColor(R.color.colorPrimaryDark));
            relation.setBackgroundColor(getResources().getColor(android.R.color.transparent));
            viewModel.setRelation(position);
        }
    }
}

File: app/src/main/java/com/example/policyfolio/ui/Nominee/AddNomineeFragment.kt
package com.example.policyfolio.ui.Nominee


import android.os.Bundle

import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProviders

import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.core.view.isGone
import androidx.core.view.isVisible

import com.example.policyfolio.R
import com.example.policyfolio.ui.adapters.ListAdapters.BasicDropdownTextAdapter
import com.example.policyfolio.util.Constants
import com.example.policyfolio.viewmodels.NomineeViewModel
import com.hbb20.CountryCodePicker
import com.wajahatkarim3.easyvalidation.core.view_ktx.nonEmpty
import com.wajahatkarim3.easyvalidation.core.view_ktx.validEmail

/**
 * A simple [Fragment] subclass.
 */
class AddNomineeFragment(private val callback: NomineeCallback) : Fragment(), BasicDropdownTextAdapter.ParentCallback {

    private var rootView: View? = null
    private var viewModel: NomineeViewModel? = null

    private var name: EditText? = null
    private var nameError: TextView? = null

    private var email: EditText? = null
    private var emailError: TextView? = null

    private var phone: LinearLayout? = null
    private var phoneText: EditText? = null
    private var ccp: CountryCodePicker? = null
    private var phoneError: TextView? = null

    private var alternativePhone: LinearLayout? = null
    private var altPhoneText: EditText? = null
    private var altCcp: CountryCodePicker? = null

    private var relationText: TextView? = null
    private var relation: LinearLayout? = null
    private var textAdapter: BasicDropdownTextAdapter? = null
    private var relations: Array<String>? = null
    private var relationError: TextView? = null

    private var done: Button? = null


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        rootView = inflater.inflate(R.layout.fragment_add_nominee, container, false)
        viewModel = ViewModelProviders.of(activity!!).get(NomineeViewModel::class.java)

        name = rootView!!.findViewById(R.id.name)
        nameError = rootView!!.findViewById(R.id.name_empty)

        email = rootView!!.findViewById(R.id.email)
        emailError = rootView!!.findViewById(R.id.email_empty)

        phoneText = rootView!!.findViewById(R.id.phone_text)
        phone = rootView!!.findViewById(R.id.phone)
        ccp = rootView!!.findViewById(R.id.ccp)
        phoneError = rootView!!.findViewById(R.id.phone_empty)


        alternativePhone = rootView!!.findViewById(R.id.alt_phone)
        altPhoneText = rootView!!.findViewById(R.id.alt_phone_text)
        altCcp = rootView!!.findViewById(R.id.ccp_alt)

        relationText = rootView!!.findViewById(R.id.relation_text)
        relation = rootView!!.findViewById(R.id.relation)
        relationError = rootView!!.findViewById(R.id.relation_empty)

        relations = resources.getStringArray(R.array.relationship_array)
        textAdapter = BasicDropdownTextAdapter(context, relations, this, Constants.ListTypes.RELATIONSHIPS, resources.getColor(R.color.colorPrimaryDarkest))

        done = rootView!!.findViewById(R.id.done)

        setListeners()

        return rootView
    }

    private fun setListeners() {
        ccp!!.registerCarrierNumberEditText(phoneText)
        ccp!!.setNumberAutoFormattingEnabled(true)

        altCcp!!.registerCarrierNumberEditText(altPhoneText)
        altCcp!!.setNumberAutoFormattingEnabled(true)

        name!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                val string = s.toString()
                if (string.isNotEmpty())
                    name!!.setBackgroundColor(resources.getColor(android.R.color.transparent))
                else
                    name!!.background = resources.getDrawable(R.drawable.text_background_grey_8dp)
            }

            override fun afterTextChanged(s: Editable) {

            }
        })

        email!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                val string = s.toString()
                if (string.isNotEmpty())
                    email!!.setBackgroundColor(resources.getColor(android.R.color.transparent))
                else
                    email!!.background = resources.getDrawable(R.drawable.text_background_grey_8dp)
            }

            override fun afterTextChanged(s: Editable) {

            }
        })

        phoneText!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                val string = s.toString()
                if (string.isNotEmpty())
                    phone!!.setBackgroundColor(resources.getColor(android.R.color.transparent))
                else
                    phone!!.background = resources.getDrawable(R.drawable.text_background_grey_8dp)
            }

            override fun afterTextChanged(s: Editable) {

            }
        })

        altPhoneText!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                val string = s.toString()
                if (string.isNotEmpty())
                    alternativePhone!!.setBackgroundColor(resources.getColor(android.R.color.transparent))
                else
                    alternativePhone!!.background = resources.getDrawable(R.drawable.text_background_grey_8dp)
            }

            override fun afterTextChanged(s: Editable) {

            }
        })

        viewModel!!.relation = -1
        relation!!.setOnClickListener { callback.openListSheet(Constants.ListTypes.RELATIONSHIPS, textAdapter) }

        done!!.setOnClickListener {
            nameError!!.isVisible = false
            nameError!!.isGone = true
            name!!.nonEmpty { message ->
                nameError!!.text = message
                nameError!!.isVisible = true
                nameError!!.isGone = false
            }

            emailError!!.isVisible = false
            emailError!!.isGone = true
            email!!.validEmail{ message ->
                emailError!!.text = message
                emailError!!.isVisible = true
                emailError!!.isGone = false
            }

            val phone = ccp!!.formattedFullNumber
            if (phone == "") {
                phoneError!!.isVisible = true
                phoneError!!.isGone = false
                phoneError!!.text = "Can't be empty!"
            } else if (!ccp!!.isValidFullNumber) {
                phoneError!!.isVisible = true
                phoneError!!.isGone = false
                phoneError!!.text = "Invalid Phone Number!"
            } else {
                phoneError!!.isVisible = false
                phoneError!!.isGone = true
            }

            if(viewModel!!.relation == -1){
                relationError!!.isVisible = true
                relationError!!.isGone = false
                relationError!!.text = "Can't be empty!"
            } else {
                relationError!!.isVisible = false
                relationError!!.isGone = true
            }

            val altPhone = altCcp!!.formattedFullNumber

            if(nameError!!.isGone && emailError!!.isGone && phoneError!!.isGone && relationError!!.isGone){
                viewModel!!.name = name!!.text.toString()
                viewModel!!.phone = phone
                viewModel!!.email = email!!.text.toString()
                if(!altPhone.isNullOrEmpty()){
                    viewModel!!.alternateNumber = altPhone
                }
                callback.done()
            }
        }
    }

    override fun setValue(position: Int, type: Int) {
        callback.closeListSheet()
        if (type == Constants.ListTypes.RELATIONSHIPS) {
            relationText!!.text = relations!![position]
            relationText!!.setTextColor(resources.getColor(R.color.colorPrimaryDark))
            relation!!.setBackgroundColor(resources.getColor(android.R.color.transparent))
            viewModel!!.relation = position
        }
    }
}
